AX: Fix to handle author error roles

hs85jeong opened the pull request:
#### 3d6388b4ceebeaf49590c8079929262246cfc100
<pre>
AX: Fix to handle author error roles
<a href="https://bugs.webkit.org/show_bug.cgi?id=261542">https://bugs.webkit.org/show_bug.cgi?id=261542</a>
rdar://problem/115462091

Reviewed by Tyler Wilcock.

This commit fixes the wai-aria/role/form-roles test[1]. According to the
ARIA spec[2], in situations where an author has not specified names for
the form and region landmarks, it is considered as authoring error. If a
valid fallback role had been specified, or if the element had an implicit
ARIA role, then user agents would continue to expose that role instead.

[1]: <a href="https://wpt.fyi/results/wai-aria/role/form-roles.html">https://wpt.fyi/results/wai-aria/role/form-roles.html</a>
[2]: <a href="https://w3c.github.io/aria/#document-handling_author-errors_roles">https://w3c.github.io/aria/#document-handling_author-errors_roles</a>

* LayoutTests/accessibility/aria-namefrom-author.html:
* LayoutTests/accessibility/roles-computedRoleString.html:
* LayoutTests/imported/w3c/web-platform-tests/wai-aria/role/form-roles-expected.txt:
* LayoutTests/platform/mac/accessibility/roles-computedRoleString-expected.txt:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAriaRoleAttribute const):

Canonical link: <a href="https://commits.webkit.org/268025@main">https://commits.webkit.org/268025@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/54ff569f8ac5fdeee174c97b0e61d0634b642e2b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18339 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18677 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19266 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20181 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17164 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18534 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21980 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18827 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19092 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18562 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18768 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15973 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21061 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15991 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16728 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23235 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17012 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16899 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21103 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17467 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14829 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16556 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/16562 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4381 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20919 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17307 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2865,11 +2865,9 @@ AccessibilityRole AccessibilityNodeObject::determineAriaRoleAttribute() const
-    // Use the native host language role of the element instead."
-    if (role == AccessibilityRole::LandmarkRegion && !hasAttribute(aria_labelAttr) && !hasAttribute(aria_labelledbyAttr))
+    // The ARIA spec states, "In situation when an author has not specified names for these landmarks such as form and region, it is considered an authoring error.
+    // If a valid fallback role had been specified, or if the element had an implicit ARIA role, then user agents would continue to expose that role, instread."
```
Might be nice to link to the spec here. https://w3c.github.io/aria/#document-handling_author-errors_roles

I also think "situation" is missing an "s", and there's a typo of "instread" vs. "instead". What do you think about using this as the comment? May need to adjust the line breaks to get it formatted more nicely 🙂 

```
// https://w3c.github.io/aria/#document-handling_author-errors_roles
// In situations where an author has not specified names for the form and
// region landmarks, it is considered an authoring error. The user agent
// MUST treat such elements as if no role had been provided.
```

hs85jeong replied:
It has become cleaner.
Thank you for the review.

twilco reviewed (approved):
Thanks again @hs85jeong! One minor nit on the comment, but looks good.

hs85jeong committed:
AX: Fix to handle author error roles
https://bugs.webkit.org/show_bug.cgi?id=261542
rdar://problem/115462091

Reviewed by Tyler Wilcock.

This commit fixes the wai-aria/role/form-roles test[1]. According to the
ARIA spec[2], in situations where an author has not specified names for
the form and region landmarks, it is considered as authoring error. If a
valid fallback role had been specified, or if the element had an implicit
ARIA role, then user agents would continue to expose that role instead.

[1]: https://wpt.fyi/results/wai-aria/role/form-roles.html
[2]: https://w3c.github.io/aria/#document-handling_author-errors_roles

* LayoutTests/accessibility/aria-namefrom-author.html:
* LayoutTests/accessibility/roles-computedRoleString.html:
* LayoutTests/imported/w3c/web-platform-tests/wai-aria/role/form-roles-expected.txt:
* LayoutTests/platform/mac/accessibility/roles-computedRoleString-expected.txt:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAriaRoleAttribute const):

Canonical link: https://commits.webkit.org/268025@main
